### PR TITLE
ppo_mujocov2

### PR DESCRIPTION
**Pull request by ShuaibinLi**

### CODE REVIEW

File: examples/PPO/README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ PARL currently supports the open-source version of Mujoco provided by DeepMind,
 + python3.7+
 + [paddle>=2.3.1](https://github.com/PaddlePaddle/Paddle)
 + [parl>=2.1.1](https://github.com/PaddlePaddle/PARL)
```
Comment (rical730) [inline examples/PPO/README.md]: please double check all package version requirements, especially parl

Comment (rical730) [inline examples/PPO/README.md]: and please update # Mujoco/Atari games introduction section in README

File: examples/PPO/README.md [hunk 2]
```diff
@@ -34,17 +34,8 @@ PARL currently supports the open-source version of Mujoco provided by DeepMind,
...
-# To train an agent for discrete action game (Atari: PongNoFrameskip-v4 by default)
-python train.py
-
-# To train an agent for continuous action game (Mujoco)
-python train.py --env 'HalfCheetah-v4' --continuous_action --train_total_steps 1000000
-```
 
-### Distributed Training
+### Training Mujoco Distributedly
 Accelerate training process by setting `xparl_addr` and `env_num > 1` when environment simulation running very slow.
```
Comment (rical730) [inline examples/PPO/README.md]: when the environment simulation runs very slowly

File: examples/PPO/README.md [hunk 3]
```diff
@@ -56,14 +47,24 @@ Note that if you have started a master before, you don't have to run the above
...
-python train.py --env 'HalfCheetah-v4' --continuous_action --train_total_steps 1000000 --env_num 5 --xparl_addr 'localhost:8010'
+
+### Training Atari
+To train an agent for discrete action game (Atari: PongNoFrameskip-v4 by default):
+
+```
+cd atari
+
+# Local training
+python train.py
```
Comment (rical730) [inline examples/PPO/README.md:69]: please add a blank line here

File: examples/PPO/mujoco/actor.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from parl.env.compat_wrappers import CompatWrapper
+
+
+@parl.remote_class(wait=False)
+class Actor(object):
+    def __init__(self, config, seed=None):
+        env = gym.make(config['env'])
+        self.env = CompatWrapper(env)
+        try:
+            self.env.seed(seed)
```
Comment (rical730) [inline examples/PPO/mujoco/actor.py]: Why we have to use `try` here, even after CompatWrapper

File: examples/PPO/mujoco/mujoco_agent.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import parl
+import paddle
+import numpy as np
+
+
+class MujocoAgent(parl.Agent):
+    def __init__(self, algorithm, config, init_beta=1.0, init_lr_multi=1.0):
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py]: Maybe `init_beta` and `init_lr_multi` can be put into config

File: examples/PPO/mujoco/mujoco_agent.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        """ Learn policy:
+
+        1. Sync parameters of policy model to old policy model
+        2. Fix old policy model, and learn policy model multi times
+        3. if use KLPEN loss, Adjust kl loss coefficient: beta
+        """
+        self.alg.sync_old_policy()
+
+        all_loss, all_kl = [], []
+        for _ in range(self.config['policy_learn_times']):
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py:92]: `policy_learn_times` looks like time related metrics. I think `policy_update_repetitions` will be better

File: examples/PPO/mujoco/mujoco_agent.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        obs = paddle.to_tensor(obs)
+        discount_sum_rewards = paddle.to_tensor(discount_sum_rewards)
+        loss = self.alg.value_learn(obs, discount_sum_rewards)
+        return loss
+
+    def policy_learn(self, obs, actions, advantages):
+        """ Learn policy:
+
+        1. Sync parameters of policy model to old policy model
+        2. Fix old policy model, and learn policy model multi times
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py]: what's the meaning of `Fix`?  please update comment

File: examples/PPO/mujoco/mujoco_agent.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        for _ in range(self.config['policy_learn_times']):
+            loss, kl, entropy = self._batch_policy_learn(obs, actions, advantages)
+            loss, kl, entropy = loss.numpy()[0], kl.numpy()[0], entropy.numpy()[0]
+            # print(self.alg.model.policy_model.state_dict()['fc_pi_std'])
+            if kl > self.kl_targ * 4:  # early stopping if D_KL diverges badly
+                break
+            all_loss.append(loss)
+            all_kl.append(kl)
+
+        # fron tf policy
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py]: please check all the comments of mujoco ppo

File: examples/PPO/mujoco/mujoco_agent.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+            obs_train = np.concatenate([obs, self.value_learn_buffer[0]])
+            discount_sum_rewards_train = np.concatenate([discount_sum_rewards, self.value_learn_buffer[1]])
+        self.value_learn_buffer = (obs, discount_sum_rewards)
+
+        all_loss = []
+        y_hat = self.alg.model.value(paddle.to_tensor(obs)).numpy().reshape(
+            [-1])  # check explained variance prior to update
+        old_exp_var = 1 - np.var(discount_sum_rewards - y_hat) / np.var(discount_sum_rewards)
+
+        for _ in range(self.config['value_learn_times']):
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py:129]: same as above, `value_update_repetitions` might be better

File: examples/PPO/mujoco/mujoco_agent.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+            while start < data_size:
+                end = start + batch_size
+                loss = self._batch_value_learn(shuffle_obs_train[start:end, :],
+                                               shuffle_discount_sum_rewards_train[start:end])
+                loss = loss.numpy()[0]
+                all_loss.append(loss)
+                start += batch_size
+        y_hat = self.alg.model.value(paddle.to_tensor(obs)).numpy().reshape(
+            [-1])  # check explained variance prior to update
+        _loss = np.mean(np.square(y_hat - discount_sum_rewards))  # explained variance after update
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py]: _loss --> value_loss

File: examples/PPO/mujoco/mujoco_agent.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        2. Fix old policy model, and learn policy model multi times
+        3. if use KLPEN loss, Adjust kl loss coefficient: beta
+        """
+        self.alg.sync_old_policy()
+
+        all_loss, all_kl = [], []
+        for _ in range(self.config['policy_learn_times']):
+            loss, kl, entropy = self._batch_policy_learn(obs, actions, advantages)
+            loss, kl, entropy = loss.numpy()[0], kl.numpy()[0], entropy.numpy()[0]
+            # print(self.alg.model.policy_model.state_dict()['fc_pi_std'])
```
Comment (rical730) [inline examples/PPO/mujoco/mujoco_agent.py]: please remove debug code

File: examples/PPO/mujoco/train.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        train_obs, train_actions, train_advantages, train_discount_sum_rewards = build_train_data(
+            config, trajectories, agent)
+
+        policy_loss, kl, beta, lr_multiplier, entropy = agent.policy_learn(train_obs, train_actions, train_advantages)
+        value_loss, exp_var, old_exp_var = agent.value_learn(train_obs, train_discount_sum_rewards)
+
+        total_train_rewards = sum([np.sum(t['rewards']) for t in trajectories])
+        logger.info('Training: Episode {}, Avg train reward: {}, Policy loss: {}, KL: {}, Value loss: {}'.format(
+            episode, total_train_rewards / len(trajectories), policy_loss, kl, value_loss))
+        summary.add_scalar("train/avg_train_reward", total_train_rewards / len(trajectories), episode)
```
Comment (rical730) [inline examples/PPO/mujoco/train.py]: avg_train_reward --> episode_mean_reward

File: parl/algorithms/paddle/ppo_mujoco.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+            model (parl.Model): model defining forward network of policy and value.
+            act_dim (float): dimension of the action space.
+            loss_type (string): loss type of PPO algorithm, 'CLIP' or 'KLPEN'".
+            kl_targ (float): D_KL target value.
+            eta (float):  multiplier for D_KL-kl_targ hinge-squared loss.
+            clip_param (float): epsilon used in the CLIP loss.
+            eps (float): A small float value for numerical stability.
+        """
+        assert isinstance(act_dim, int)
+        assert isinstance(clip_param, float)
```
Comment (rical730) [inline parl/algorithms/paddle/ppo_mujoco.py:49]: Please verify the methods and member variables that `model` needs to implement

File: examples/PPO/README.md [hunk 14]
```diff
@@ -24,7 +24,7 @@ PARL currently supports the open-source version of Mujoco provided by DeepMind,
 + python3.7+
```
Comment (rical730) [inline examples/PPO/README.md:26]: Please explain that the horizontal axis of Mujoco results is episode

File: examples/PPO/README.md (hunk 1 above)
Comment (ShuaibinLi) [reply examples/PPO/README.md]: done
